src/GraphQL.Client/GraphQLHttpClient.cs

```diff
@@ -111,6 +116,62 @@ public IObservable<GraphQLResponse<TResponse>> CreateSubscriptionStream<TRespons
 
     #region Private Methods
 
+    private async Task<GraphQLHttpResponse<TResponse>> SendAPQHttpRequestAsync<TResponse>(GraphQLRequest request, CancellationToken cancellationToken = default)
+    {
+        cancellationToken.ThrowIfCancellationRequested();
+
+        var savedQuery = request.Query;
+        bool useAPQ = false;
+
+        if (request.Query != null && _useAPQ && Options.EnableAutomaticPersistedQueries(request))
+        {
+            // https://www.apollographql.com/docs/react/api/link/persisted-queries/
+            const int APQ_SUPPORTED_VERSION = 1;
+            useAPQ = true;
+            // TODO: I suggest to change request.Extension type to public Dictionary<string, object?>?
+            //request.Extensions ??= new();
+            //request.Extensions["persistedQuery"] = new Dictionary<string, object>
+            //{
+            //    ["version"] = APQ_SUPPORTED_VERSION,
+            //    ["sha256Hash"] = Hash.Compute(request.Query),
+            //};
+            request.Extensions ??= new
+            {
+                version = APQ_SUPPORTED_VERSION,
+                sha256Hash = Hash.Compute(request.Query),
+            };
+            request.Query = null;
+        }
+
+        var response = await SendHttpRequestAsync<TResponse>(request, cancellationToken);
+
+        if (useAPQ && response.Errors?.Length > 0)
+        {
+            if (response.Errors.Any(error => string.Equals(error.Message, "PersistedQueryNotFound", StringComparison.CurrentCultureIgnoreCase)))
+            {
+                // GraphQL server supports APQ!
+
+                // Alas, for the first time we did not guess and in vain removed Query, so we return Query and
+                // send request again. This is one-time "cache miss", not so scary. By the way, after a second call,
+                // there is a probability of receiving PERSISTED_QUERY_NOT_FOUND error again, but it is so ghostly
+                // that this can and should be neglect so not to complicate the code without the need.
+                request.Query = savedQuery;
+                return await SendHttpRequestAsync<TResponse>(request, cancellationToken);
+            }
+            else if (response.Errors.Any(error => string.Equals(error.Message, "PersistedQueryNotSupported") ||
+                     response.Errors.Any(error => string.Equals(error.Message, "GraphQL query is missing.")))) // GraphQL.NET specific error message
```

Comment:
What if it's another GraphQL server which has a slightly different response?  This repo is not exclusively for GraphQL servers that either (1) support APQ or (2) are GraphQL.NET.  I'm not sure, but even GraphQL.NET might return different messages for older versions.

Reply:
I know. It's more like demo to discuss. My implementation has more conditions to consider - http status codes, graphql error codes in extensions, etc. - but I minimized them to minimum to make code clean. I'm open to suggestions.

Reply:
Apollo uses a delegate to determine if it should disable persistent queries for the session... 

> `disable`: a function which takes an `ErrorResponse` (see below) and returns a boolean to disable any future persisted queries for that session. This defaults to disabling on `PersistedQueryNotSupported` or a 400 or 500 http error.

We'd have to handle the `GraphQLHttpRequestException` from `SendHttpRequestAsync` there, too.

Reply:
the error handling is a bit of a pain, there doesn't seem to be any error code or structured response indicating missing or incorrect hash, I'm afraid we must just link currently known responses, for particular implementations to behaviors we want. Maybe at one point they will come up with a generic way to report such errors.

Reply:
> We'd have to handle the GraphQLHttpRequestException from SendHttpRequestAsync there, too

Still in TODO.